File: glitch/backend/app/schema/item.py
from pydantic import BaseModel  # type: ignore
from typing import List, Optional


class Item(BaseModel):
    rid: int
    id_project: int
    type: int
    state: int
    risk: int
    risk_factors: int
    priority: int
    title: str
    detail: str
    result: str
    datetime_entry: str
    datetime_update: str
    rid_users: int
    name: Optional[str] = None
    rid_users_review: Optional[int] = None
    name_review: Optional[str] = None
    project_datetime_start: Optional[str] = None
    project_datetime_end: Optional[str] = None
    event_datetime_end: Optional[str] = None
    story_datetime_start: Optional[str] = None
    story_datetime_end: Optional[str] = None
    task_type: Optional[int] = None
    task_workload: Optional[int] = None
    task_number_completed: Optional[int] = None
    task_number_total: Optional[int] = None
    bug_workload: Optional[int] = None

    class Config:
        orm_mode = True


class Project(BaseModel):
    rid: int
    id_project: int
    state: int
    risk: int
    risk_factors: int
    title: str
    detail: str
    result: str
    datetime_entry: str
    datetime_update: str
    rid_users: int
    name: Optional[str] = None
    project_datetime_start: Optional[str] = None
    project_datetime_end: Optional[str] = None

    class Config:
        orm_mode = True


class ProjectCreate(BaseModel):
    rid_users: int
    title: str
    detail: str
    datetime_start: str
    datetime_end: str

    class Config:
        orm_mode = True


class ProjectUpdate(BaseModel):
    rid: int
    state: int
    rid_users: int
    rid_users_review: Optional[int] = None
    title: str
    detail: str
    result: str
    datetime_start: str
    datetime_end: str

    class Config:
        orm_mode = True


class EventCreate(BaseModel):
    id_project: int
    rid_items: int
    rid_users: int
    title: str
    detail: str
    datetime_end: str

    class Config:
        orm_mode = True


class EventUpdate(BaseModel):
    rid: int
    state: int
    rid_users: int
    rid_users_review: Optional[int] = None
    title: str
    detail: str
    result: str
    datetime_end: str

    class Config:
        orm_mode = True


class FeatureCreate(BaseModel):
    id_project: int
    rid_items: int
    rid_users: int
    title: str
    detail: str

    class Config:
        orm_mode = True


class FeatureUpdate(BaseModel):
    rid: int
    state: int
    rid_users: int
    rid_users_review: Optional[int] = None
    title: str
    detail: str
    result: str

    class Config:
        orm_mode = True


class StoryCreate(BaseModel):
    id_project: int
    rid_items: int
    rid_users: int
    title: str
    detail: str
    datetime_start: str
    datetime_end: str

    class Config:
        orm_mode = True


class StoryUpdate(BaseModel):
    rid: int
    state: int
    rid_users: int
    rid_users_review: Optional[int] = None
    title: str
    detail: str
    result: str
    datetime_start: str
    datetime_end: str

    class Config:
        orm_mode = True


class TaskCreate(BaseModel):
    id_project: int
    rid_items: int
    rid_users: int
    title: str
    detail: str
    type: int
    workload: int
    number_completed: int
    number_total: int

    class Config:
        orm_mode = True


class TaskUpdate(BaseModel):
    rid: int
    state: int
    rid_users: int
    rid_users_review: Optional[int] = None
    title: str
    detail: str
    result: str
    type: int
    workload: int
    number_completed: int
    number_total: int

    class Config:
        orm_mode = True


class TaskPriorityUpdate(BaseModel):
    rid: int
    priority: int

    class Config:
        orm_mode = True


class BugCreate(BaseModel):
    id_project: int
    rid_items: int
    rid_users: int
    title: str
    detail: str
    workload: int

    class Config:
        orm_mode = True


class BugUpdate(BaseModel):
    rid: int
    state: int
    rid_users: int
    rid_users_review: Optional[int] = None
    title: str
    detail: str
    result: str
    workload: int

    class Config:
        orm_mode = True


class BugPriorityUpdate(BaseModel):
    rid: int
    priority: int

    class Config:
        orm_mode = True


class ItemHierarchy(BaseModel):
    rid: int
    rid_users: int
    name: str
    type: int
    title: str
    workload_task: Optional[int] = None
    workload_bug: Optional[int] = None
    children: Optional[List['ItemHierarchy']] = None

    class Config:
        orm_mode = True 

class SummaryProject(BaseModel):
    rid: int
    type: int
    state: int
    title: str
    datetime_start: str
    datetime_end: str

    class Config:
        orm_mode = True 

File: glitch/frontend/app/glitch/src/types/Item.ts
export enum ItemType {
  NONE = 0,
  PROJECT,
  EVENT,
  FEATURE,
  STORY,
  TASK,
  BUG
}

export enum ItemState {
  NONE = 0,
  IDLE,
  RUN,
  ALERT,
  REVIEW,
  COMPLETE
}

export enum TaskType {
  NONE = 0,
  WORKLOAD,
  NUMBER
}

export const RISK_1 = 0b00000001
export const RISK_2 = 0b00000010
export const RISK_3 = 0b00000100
export const RISK_4 = 0b00001000
export const RISK_5 = 0b00010000
export const RISK_6 = 0b00100000
export const RISK_7 = 0b01000000
export const RISK_8 = 0b10000000

export enum WorkloadType {
  NONE = 0,
  WITHIN_AN_HOUR = 1,
  WITHIN_HALF_A_DAY = 3,
  WITHIN_A_DAY = 7,
  WITHIN_2_DAYS = 14,
  WITHIN_3_DAYS = 21,
  WITHIN_A_WEEK = 35
}

export enum ExtractType {
  NONE = 0,
  ALL,
  INCOMPLETE,
  HIGH_RISK,
  ALERT,
  ASSIGNMENT,
  RELATION,
  SEARCH
}

export interface Item {
  rid: number
  id_project: number
  type: ItemType
  state: ItemState
  risk: number
  risk_factors: number
  priority: number
  title: string
  detail: string
  result: string
  datetime_entry: string
  datetime_update: string
  rid_users: number
  name: string
  rid_users_review: number | null
  name_review: string | null
  project_datetime_start: string
  project_datetime_end: string
  event_datetime_end: string
  story_datetime_start: string
  story_datetime_end: string
  task_type: TaskType
  task_workload: number
  task_number_completed: number
  task_number_total: number
  bug_workload: number
}

export interface Project {
  rid: number
  id_project: number
  state: number
  risk: number
  title: string
  detail: string
  result: string
  datetime_entry: string
  datetime_update: string
  rid_users: number
  name: string
  project_datetime_start: string
  project_datetime_end: string
}

export interface ProjectCreate {
  rid_users: number
  title: string
  detail: string
  datetime_start: string
  datetime_end: string
}

export interface ProjectUpdate {
  rid: number
  state: number
  rid_users: number
  rid_users_review: number | null
  title: string
  detail: string
  result: string
  datetime_start: string
  datetime_end: string
}

export interface EventCreate {
  id_project: number
  rid_items: number
  rid_users: number
  title: string
  detail: string
  datetime_end: string
}

export interface EventUpdate {
  rid: number
  state: number
  rid_users: number
  rid_users_review: number | null
  title: string
  detail: string
  result: string
  datetime_end: string
}

export interface FeatureCreate {
  id_project: number
  rid_items: number
  rid_users: number
  title: string
  detail: string
}

export interface FeatureUpdate {
  rid: number
  state: number
  rid_users: number
  rid_users_review: number | null
  title: string
  detail: string
  result: string
}

export interface StoryCreate {
  id_project: number
  rid_items: number
  rid_users: number
  title: string
  detail: string
  datetime_start: string
  datetime_end: string
}

export interface StoryUpdate {
  rid: number
  state: number
  rid_users: number
  rid_users_review: number | null
  title: string
  detail: string
  result: string
  datetime_start: string
  datetime_end: string
}

export interface TaskCreate {
  id_project: number
  rid_items: number
  rid_users: number
  title: string
  detail: string
  type: number
  workload: number
  number_completed: number
  number_total: number
}

export interface TaskUpdate {
  rid: number
  state: number
  rid_users: number
  rid_users_review: number | null
  title: string
  detail: string
  result: string
  type: number
  workload: number
  number_completed: number
  number_total: number
}

export interface TaskPriorityUpdate {
  rid: number
  priority: number
}

export interface BugCreate {
  id_project: number
  rid_items: number
  rid_users: number
  title: string
  detail: string
  workload: number
}

export interface BugUpdate {
  rid: number
  state: number
  rid_users: number
  rid_users_review: number | null
  title: string
  detail: string
  result: string
  workload: number
}

export interface BugPriorityUpdate {
  rid: number
  priority: number
}

export interface ItemHierarchy {
  rid: number
  rid_users: number
  name: string
  title: string
  workload_task?: number
  workload_bug?: number
  children?: ItemHierarchy[]
}

export interface SummaryProject {
  rid: number
  type: number
  state: number
  title: string
  datetime_start: string
  datetime_end: string
}

export interface ItemFrequency {
  datetime_entry: string
  task_count: number
  task_count_idle: number
  task_count_run: number
  task_count_alert: number
  task_count_review: number
  task_count_complete: number
  bug_count: number
  bug_count_idle: number
  bug_count_run: number
  bug_count_alert: number
  bug_count_review: number
  bug_count_complete: number
}
